src/iop/hazeremoval.c

```diff
@@ -0,0 +1,962 @@
+{
+  free(self->params);
+  self->params = NULL;
+}
+
+static void strength_callback(GtkWidget *w, dt_iop_module_t *self)
+{
+  if (self->dt->gui->reset)
+    return;
+  dt_iop_hazeremoval_params_t *p = (dt_iop_hazeremoval_params_t *)self->params;
+  p->strength = dt_bauhaus_slider_get(w);
+  dt_dev_add_history_item(darktable.develop, self, TRUE);
+}
+
+static void distance_callback(GtkWidget *w, dt_iop_module_t *self)
+{
+  if (self->dt->gui->reset)
+    return;
+  dt_iop_hazeremoval_params_t *p = (dt_iop_hazeremoval_params_t *)self->params;
+  p->distance = dt_bauhaus_slider_get(w);
+  dt_dev_add_history_item(darktable.develop, self, TRUE);
+}
+
+void gui_update(struct dt_iop_module_t *self)
+{
+  dt_iop_module_t *module = (dt_iop_module_t *)self;
+  dt_iop_hazeremoval_gui_data_t *g = (dt_iop_hazeremoval_gui_data_t *)self->gui_data;
+  dt_iop_hazeremoval_params_t *p = (dt_iop_hazeremoval_params_t *)module->params;
+  dt_bauhaus_slider_set(g->strength, p->strength);
+  dt_bauhaus_slider_set(g->distance, p->distance);
+
+  dt_pthread_mutex_lock(&g->lock);
+  g->distance_max = NAN;
+  g->A0[0] = NAN;
+  g->A0[1] = NAN;
+  g->A0[2] = NAN;
+  g->hash = 0;
+  dt_pthread_mutex_unlock(&g->lock);
+}
+
+void gui_init(dt_iop_module_t *self)
+{
+  self->gui_data = malloc(sizeof(dt_iop_hazeremoval_gui_data_t));
+  dt_iop_hazeremoval_gui_data_t *g = (dt_iop_hazeremoval_gui_data_t *)self->gui_data;
+  dt_iop_hazeremoval_params_t *p = (dt_iop_hazeremoval_params_t *)self->params;
+
+  dt_pthread_mutex_init(&g->lock, NULL);
+  g->distance_max = NAN;
+  g->A0[0] = NAN;
+  g->A0[1] = NAN;
+  g->A0[2] = NAN;
+  g->hash = 0;
+
+  self->widget = gtk_box_new(GTK_ORIENTATION_VERTICAL, DT_BAUHAUS_SPACE);
+
+  g->strength = dt_bauhaus_slider_new_with_range(self, -1, 1, 0.01, p->strength, 2);
+  dt_bauhaus_widget_set_label(g->strength, NULL, _("strength"));
+  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(g->strength), TRUE, TRUE, 0);
+  g_signal_connect(G_OBJECT(g->strength), "value-changed", G_CALLBACK(strength_callback), self);
+
+  g->distance = dt_bauhaus_slider_new_with_range(self, 0, 1, 0.005, p->distance, 3);
+  dt_bauhaus_widget_set_label(g->distance, NULL, _("distance"));
+  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(g->distance), TRUE, TRUE, 0);
+  g_signal_connect(G_OBJECT(g->distance), "value-changed", G_CALLBACK(distance_callback), self);
+}
+
+void gui_cleanup(dt_iop_module_t *self)
+{
+  dt_iop_hazeremoval_gui_data_t *g = (dt_iop_hazeremoval_gui_data_t *)self->gui_data;
+  dt_pthread_mutex_destroy(&g->lock);
+  free(self->gui_data);
+  self->gui_data = NULL;
+}
+
+//----------------------------------------------------------------------
+// module local functions and structures required by process function
+//----------------------------------------------------------------------
+
+typedef struct tile {
+  int left, right, lower, upper;
+} tile;
+
+typedef struct rgb_image {
+  rgb_pixel *data;
+  int width, height, stride;
```

Comment:
ah wait, stride is count of channels?
why introduce yet another name, if ch is used everywhere?